test/src/test/groovy/org/openremote/test/protocol/mqtt/TeltonikaMQTTProtocolTest.groovy

```diff
@@ -0,0 +1,284 @@
+import spock.lang.Specification
+import spock.util.concurrent.PollingConditions
+
+class TeltonikaMQTTProtocolTest extends Specification implements ManagerContainerTrait {
+    @Shared public def conditions = new PollingConditions(timeout: 5, delay: 0.2)
+
+    public static Container container
+    public static AssetStorageService assetStorageService;
+    public static AgentService agentService;
+    public static MQTTBrokerService mqttBrokerService
+    public static ManagerTestSetup managerTestSetup;
+    public static KeycloakTestSetup keycloakTestSetup;
+    public static ClientEventService clientEventService;
+    public static String mqttHost;
+    public static int mqttPort;
+    public static String mqttClientId;
+    public static String username;
+    public static String password;
+    public static MQTT_IOClient client;
+    public static TeltonikaMQTTHandler handler;
+
+    //TODO: I do not know how/if to link this to environment variable
+    String TELTONIKA_DEVICE_RECEIVE_TOPIC = "data"
+    String TELTONIKA_DEVICE_SEND_TOPIC = "commands";
+    String TELTONIKA_DEVICE_TOKEN = "teltonika";
+    String TELTONIKA_DEVICE_SEND_COMMAND_ATTRIBUTE_NAME = "sendToDevice";
+    String TELTONIKA_DEVICE_RECEIVE_COMMAND_ATTRIBUTE_NAME = "response";
+    //Real IMEI: https://www.imei.info/?imei=358491098808487
+    def TELTONIKA_DEVICE_IMEI = "358491098808487";
+
+    def setupSpec() {
+        and: "the container is started"
+        container = startContainer(defaultConfig(), defaultServices());
+        assetStorageService = container.getService(AssetStorageService.class)
+        agentService = container.getService(AgentService.class)
+        mqttBrokerService = container.getService(MQTTBrokerService.class)
+        managerTestSetup = container.getService(SetupService.class).getTaskOfType(ManagerTestSetup.class)
+        keycloakTestSetup = container.getService(SetupService.class).getTaskOfType(KeycloakTestSetup.class)
+        clientEventService = container.getService(ClientEventService.class)
+        handler = mqttBrokerService.customHandlers.find {it instanceof TeltonikaMQTTHandler} as TeltonikaMQTTHandler
+
+        mqttHost = mqttBrokerService.host
+        mqttPort = mqttBrokerService.port
+
+        mqttClientId = UniqueIdentifierGenerator.generateId()
+        username = keycloakTestSetup.realmBuilding.name + ":" + keycloakTestSetup.serviceUser.username // realm and OAuth client id
+        password = keycloakTestSetup.serviceUser.secret
+
+
+        client = new MQTT_IOClient(mqttClientId, mqttHost, mqttPort, false, false,  null, null, null)
+    }
+
+    def setup(){
+//        I think that it is required to provide usernamePassword to connect the client to the broker. Look at AbstractMQTT_IOClient:L100.
+        client.removeAllMessageConsumers();
+        client.disconnect();
+
+        client = new MQTT_IOClient(mqttClientId, mqttHost, mqttPort, false, false, null, null, null)
+    }
+
+    def cleanupSpec(){
```

Comment:
I think cleanup methods shouldn't be necessary. See `MqttBrokerTest` for inspiration.

Reply:
I think they are because I use multiple tests for my groovy tests, and they just help with keeping things neat I believe.